src/Init/Data/String/Basic.lean

```diff
@@ -290,73 +290,111 @@ where go (acc : String) (s : String) : List String → String
   | a :: as => go (acc ++ s ++ a) s as
   | []      => acc
 
-/-- Iterator for `String`. That is, a `String` and a position in that string. -/
+/-- Iterator over the characters (`Char`) of a `String`.
+
+Typically created by `s.iter`, where `s` is a `String`. -/
 structure Iterator where
+  /-- The string the iterator is for. -/
   s : String
+  /-- The current position.
+
+  This position is not necessarily legal for the string, for instance if one keeps calling
+  `Iterator.next` when `Iterator.atEnd` is true. If the position is not a legal one, then the
+  current character is `(default : Char)`, similar to `String.get` on an illegal position. -/
```

Comment:
I think the wording here should discourage invalid positions a bit more strongly. There is no proof witness in the type asserting that the position is valid, but it should be unequivocally a logic bug to have such a misaligned iterator and all string functions should be allowed to have arbitrary behavior in such cases.

Reply:
I understand what you're saying but I find this a bit tricky. I think one expects "position `i` is valid for string `s`" to mean that `i` is in-bounds, but then `"".iter` is invalid, and so is `"1".iter.next` although `"1".iter.atEnd` is false.

So I assume you mean that `i` is valid for `s` if `i` is in-bounds or `i` is the successor of the index of the last character in `s`. Is that what you mean?

Reply:
A `p : String.Pos` is valid for `s : String` when `0 <= p <= s.endPos` and `p` lies on a UTF8 byte boundary. This is defined in std.

Reply:
Yes, there are two options of validity here: valid for character access, and valid for iterator comparison (which also allows `atEnd`).

Maybe the C++ documentation has some terminology for this distinction that we can copy.

Reply:
In C++ a pointer is considered valid (not immediate UB) if it is within the allocation or one-past-the-end. I'd rather not borrow the one-past-the-end terminology though, it's very awkward and reinforces the idea that a pointer points to an element rather than a separation point "between characters". Std currently does not have any terminology for "valid and not at the end"; if necessary it would just say that as two things. [`String.get_of_valid`](https://leanprover-community.github.io/mathlib4_docs/find/?pattern=String.get_of_valid#src) is not stated in a way that directly uses `String.Pos.Valid` because it needs to know more specifically what is the list of characters on the right of the partition.

Reply:
How about this, with the exceptions mentioned properly documented?

```lean
/--

An iterator is *valid* if the position `i` is valid for the string `s`, meaning `0 ≤ i ≤ s.endPos`
and `i` lies on a UTF8 by boundary.

Most operations on iterators return arbitrary values if the iterator is not valid. The functions in
the `String.Iterator` API should prevent you from dealing with invalid iterators, with two
exceptions:

- `Iterator.next iter` is invalid if `iter` is already at the end of the string (`iter.atEnd` is
  true), and
- `Iterator.forward iter n`/`Iterator.nextn iter n` if `n` is strictly greater than the number of
  remaining characters. -/
```

Reply:
The definition of string position validity should also be moved or copied to the `String.Pos` docs.

Reply:
Done

Reply:
Done

Reply:
I think there's no need to emphasize `*at the end of the string*` here - this is an ordinary and expected state for an iterator to have.

How about this minor edit:
```
An iterator is *valid* if its current position `i` is valid for the string `s`, meaning `0 ≤ i ≤ s.endPos`
and `i` lies on a UTF8 byte boundary. If `i = s.endPos`, the iterator is at the end of the string.

Most operations on iterators return arbitrary values if the iterator is not valid. The functions in
the `String.Iterator` API should rule out the creation of invalid iterators, with two
exceptions:

- `Iterator.next iter` is invalid if `iter` is already at the end of the string (`iter.atEnd` is
  true), and
- `Iterator.forward iter n`/`Iterator.nextn iter n` if `n` is strictly greater than the number of
  remaining characters.
```

Reply:
Done, with the link to std4 removed